File: composeApp/src/commonMain/kotlin/kr/hhp227/storygroup/ui/screens/group/GroupReportsViewModel.kt
package kr.hhp227.storygroup.ui.screens.group

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kr.hhp227.storygroup.shared.domain.model.PostReport
import kr.hhp227.storygroup.shared.domain.model.ReportStatus
import kr.hhp227.storygroup.shared.domain.usecase.GetGroupReportsUseCase
import kr.hhp227.storygroup.shared.domain.usecase.ProcessGroupReportUseCase
import kr.hhp227.storygroup.ui.mvi.MviViewModel

/**
 * 그룹 신고함(모더레이터 전용) — 웹 /groups/[id]/reports 미러. 필터(대기중/전체)와
 * 확인/기각 처리. 처리 성공은 대기중 필터에선 행 제거, 전체에선 상태만 갱신(웹과 동일).
 * 화면 전환(게시글 상세)은 화면 콜백 직행이라 EVENT는 Nothing.
 * iosApp GroupReportsViewModel.swift와 1:1 미러
 */
class GroupReportsViewModel(
    val groupId: Long,
    private val getGroupReportsUseCase: GetGroupReportsUseCase,
    private val processGroupReportUseCase: ProcessGroupReportUseCase
) : ViewModel(), MviViewModel<GroupReportsViewModel.UiState, GroupReportsViewModel.Action, Nothing> {
    private val _uiState = MutableStateFlow(UiState())
    override val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    override val event: Flow<Nothing> = emptyFlow()

    override fun onAction(action: Action) {
        when (action) {
            Action.Refresh -> refresh()
            is Action.SetFilter -> setFilter(action.filter)
            is Action.Process -> process(action.reportId, action.status)
            Action.DismissActionError -> _uiState.update { it.copy(actionError = null) }
        }
    }

    private fun refresh() {
        if (_uiState.value.isLoading) return

        _uiState.update { it.copy(isLoading = true, loadError = null) }
        viewModelScope.launch {
            val filter = _uiState.value.filter

            runCatching { getGroupReportsUseCase(groupId, filter) }
                .onSuccess { reports ->
                    _uiState.update { it.copy(isLoading = false, reports = reports) }
                }
                .onFailure { e ->
                    _uiState.update {
                        it.copy(isLoading = false, loadError = e.message ?: "신고 목록을 불러오지 못했습니다.")
                    }
                }
        }
    }

    /** 필터 전환 — 목록을 비우고 다시 읽는다(웹의 key 리마운트 미러) */
    private fun setFilter(filter: ReportStatus?) {
        if (_uiState.value.filter == filter) return

        _uiState.update { it.copy(filter = filter, reports = null, isLoading = false, loadError = null) }
        refresh()
    }

    private fun process(reportId: Long, status: ReportStatus) {
        if (_uiState.value.busyReportId != null) return

        _uiState.update { it.copy(busyReportId = reportId, actionError = null) }
        viewModelScope.launch {
            runCatching { processGroupReportUseCase(groupId, reportId, status) }
                .onSuccess { updated ->
                    _uiState.update { state ->
                        // 대기중 필터에서는 처리된 행이 빠지고, 전체에서는 상태만 갱신된다(웹 미러)
                        val reports = state.reports?.mapNotNull { report ->
                            when {
                                report.id != reportId -> report
                                state.filter == ReportStatus.PENDING -> null
                                else -> updated
                            }
                        }

                        state.copy(busyReportId = null, reports = reports)
                    }
                }
                .onFailure { e ->
                    _uiState.update { it.copy(busyReportId = null, actionError = e.message ?: "신고 처리에 실패했습니다.") }
                }
        }
    }

    init {
        refresh()
    }

    data class UiState(
        // null=아직 로드 전 — 빈 목록(신고 없음)과 구분한다
        val reports: List<PostReport>? = null,
        val isLoading: Boolean = false,
        val loadError: String? = null,
        // 기본 대기중 — null이면 전체(웹 필터 칩 미러)
        val filter: ReportStatus? = ReportStatus.PENDING,
        // 처리 진행 중인 행 — 그 행의 버튼만 잠근다(웹 busyFor 미러)
        val busyReportId: Long? = null,
        val actionError: String? = null
    )

    sealed interface Action {
        data object Refresh : Action
        data class SetFilter(val filter: ReportStatus?) : Action
        data class Process(val reportId: Long, val status: ReportStatus) : Action
        data object DismissActionError : Action
    }
}

File: iosApp/iosApp/UI/Screens/Group/GroupReportsViewModel.swift
import Combine
import Foundation
import Shared

/// 그룹 신고함(모더레이터 전용) — composeApp GroupReportsViewModel.kt와 1:1 미러
/// (웹 /groups/[id]/reports). 필터(대기중/전체)와 확인/기각 처리. 처리 성공은 대기중
/// 필터에선 행 제거, 전체에선 상태만 갱신(웹과 동일). 화면 전환은 뷰 직행이라 Event=Never.
final class GroupReportsViewModel: MviViewModel {
    typealias Event = Never

    @Published private(set) var uiState = UiState()

    let groupId: Int64

    private let getGroupReportsUseCase: GetGroupReportsUseCase

    private let processGroupReportUseCase: ProcessGroupReportUseCase

    func onAction(_ action: Action) {
        switch action {
        case .refresh: refresh()
        case .setFilter(let filter): setFilter(filter)
        case .process(let reportId, let status): process(reportId, status)
        case .dismissActionError: uiState.actionError = nil
        }
    }

    private func refresh() {
        if uiState.isLoading { return }

        uiState.isLoading = true
        uiState.loadError = nil
        Task { @MainActor in
            do {
                let reports = try await getGroupReportsUseCase.invoke(groupId: groupId, status: uiState.filter)
                uiState.isLoading = false
                uiState.reports = reports
            } catch {
                uiState.isLoading = false
                uiState.loadError = error.kotlinMessage(fallback: "신고 목록을 불러오지 못했습니다.")
            }
        }
    }

    /// 필터 전환 — 목록을 비우고 다시 읽는다(웹의 key 리마운트 미러)
    private func setFilter(_ filter: ReportStatus?) {
        if uiState.filter == filter { return }

        uiState.filter = filter
        uiState.reports = nil
        uiState.isLoading = false
        uiState.loadError = nil
        refresh()
    }

    private func process(_ reportId: Int64, _ status: ReportStatus) {
        if uiState.busyReportId != nil { return }

        uiState.busyReportId = reportId
        uiState.actionError = nil
        Task { @MainActor in
            do {
                let updated = try await processGroupReportUseCase.invoke(groupId: groupId, reportId: reportId, status: status)
                uiState.busyReportId = nil
                // 대기중 필터에서는 처리된 행이 빠지고, 전체에서는 상태만 갱신된다(웹 미러)
                uiState.reports = uiState.reports?.compactMap { report in
                    if report.id != reportId { return report }
                    return uiState.filter == .pending ? nil : updated
                }
            } catch {
                uiState.busyReportId = nil
                uiState.actionError = error.kotlinMessage(fallback: "신고 처리에 실패했습니다.")
            }
        }
    }

    init(groupId: Int64, getGroupReportsUseCase: GetGroupReportsUseCase, processGroupReportUseCase: ProcessGroupReportUseCase) {
        self.groupId = groupId
        self.getGroupReportsUseCase = getGroupReportsUseCase
        self.processGroupReportUseCase = processGroupReportUseCase
        refresh()
    }

    struct UiState {
        // nil=아직 로드 전 — 빈 목록(신고 없음)과 구분한다
        var reports: [PostReport]? = nil
        var isLoading = false
        var loadError: String? = nil
        // 기본 대기중 — nil이면 전체(웹 필터 칩 미러)
        var filter: ReportStatus? = .pending
        // 처리 진행 중인 행 — 그 행의 버튼만 잠근다(웹 busyFor 미러)
        var busyReportId: Int64? = nil
        var actionError: String? = nil
    }

    enum Action {
        case refresh
        case setFilter(ReportStatus?)
        case process(Int64, ReportStatus)
        case dismissActionError
    }
}
